### ISSUE TITLE
3D/Volumetric Convolutional Neural Networks support

### ISSUE THREAD

**Issue by archenroot** [deeplearning4j/deeplearning4j#3174]
#### Issue Description

Is there any kind of support for 3D/Volumetric Convolutional Neural Networks in deeplearning4j? If yes, any example available?

Something like:
https://github.com/dimatura/voxnet

**Comment by AlexDBlack** [comment]
No 3d CNN support at present. We don't currently have 3d CNN layers, plus we would also need to add some supporting functionality (such as preprocessors to handle 3d CNN -> dense, etc; plus dataset merging, etc) for this case.
It's not on the immediate roadmap for us at present.

**Comment by agibsonccc** [comment]
https://github.com/deeplearning4j/deeplearning4j/issues/3053 Flagging this as well.

**Comment by turambar** [comment]
Overlaps https://github.com/deeplearning4j/deeplearning4j/issues/2447. Should we consolidate and/or close one of these?

**Comment by agibsonccc** [comment]
Close it in favor of this one. Reason being: we have to implement 3d convolutions first. We can do keras after that.

**Comment by jadedfire** [comment]
looking at the code, it appears that assumptions concerning multidimensional shape were made all the way down into ND4J as i find "row" and "column" all over the place (i.e. always assume 2D). i suspect 3D support will require a major overhaul if my limited research is accurate.

**Comment by agibsonccc** [comment]
@jadedfire could you clarify? Nd4j itself is a tensor library we have ndarrays everywhere (you don't have to look far - even our current cnn implementation is 4d o_0)

**Comment by raver119** [comment]
Fortunately your research isn't accurate. Nd4j supports up to 32 dimensions :)

**Comment by jadedfire** [comment]
my apologies, i will dig deeper

**Comment by archenroot** [comment]
I am from Czech republic and there is widely used search engine is www.seznam.cz, which provides as well now 3D maps view generated by engine which is developed by https://www.melown.com. In general this engine is open source. I know it is not Java, but in general this could be component which will take any image and convert it into 3d model... I am not sure about all the capabilities in the moment, but maybe worth to investigate available mechanisms and possible compatibility bridge with deeplearning4j.

https://www.melown.com
https://github.com/Melown

**Comment by agibsonccc** [comment]
@archenroot We'll implement it when we get time. We usually don't expect external contributions for layers. The barrier is a bit high for 99.9% of folks out there ;/

**Comment by archenroot** [comment]
@agibsonccc I see, I just red about them and see that they produce engine for transforming images into 3D models, so just got idea about some integration, etc. But in general I understand that it will be more consistent to have it fully integrated inside deeplearning stack rather than reusing some "unknown" technology.

**Comment by archenroot** [comment]
Maybe this could be possible project to adopt: https://github.com/chrischoy/3D-R2N2/

**Comment by archenroot** [comment]
@agibsonccc may I ask you if you can provide any estimation of this feature request to be implemented? even gross estimation is ok, thx.

**Comment by agibsonccc** [comment]
No. If skymind doesnt need it, its not being built. We only build things
for customers.

**Comment by archenroot** [comment]
@agibsonccc I see and understand, is there any docs, or example to follow how I can try to adapt/implement it by myself?

**Comment by turambar** [comment]
Reminder: we have to implement this in DL4J before we can import it from Keras so perhaps not @maxpumperla's purview?

**Comment by raver119** [comment]
Implemented in libnd4j quite a while ago.

**Comment by maxpumperla** [comment]
@raver119 there's an open PR for this here #4927. let's not just close things like this.
